cmd/controller-manager/app/controllermanager.go

```diff
@@ -181,6 +187,17 @@ func startClusterController(ctx controllerscontext.Context) (enabled bool, err e
 	if err := clusterController.SetupWithManager(mgr); err != nil {
 		return false, err
 	}
+
+	taintManager := &cluster.NoExecuteTaintManager{
+		Client:                             mgr.GetClient(),
+		EventRecorder:                      mgr.GetEventRecorderFor(cluster.TaintManagerName),
+		ClusterTaintEvictionRetryFrequency: 10 * time.Second,
+		ConcurrentReconciles:               3,
+	}
+	if err := taintManager.SetupWithManager(mgr); err != nil {
```

Comment:
The NoExecuteTaintManager is essentially a controller, do you think we should reserve the capacity of disabling it?

Reply:
I think NoExecuteTaintManager is a part of cluster controller. 

Kubernetes takes `TaintManager` as a part of NodeLifecycleController. It provides an option `--enable-taint-manager` in `NodeLifecycleController` to enable taint manager.

So what about adding an option `--enable-taint-manager` as well?

Reply:
I agree. the --enable-taint-manager defaults to `true`. But, I think the feature is part of `Failover`, I suggest checking both the [feature gate](https://github.com/karmada-io/karmada/blob/f46d90f6b4ab04e5263aa142c3b6ae6078b645da/pkg/features/features.go#L10) and the flags. 
By the way, the feature gate will be removed eventually.

Reply:
> But, I think the feature is part of Failover.

Actually, I think this feature is independent. We implement NoExecute taint manager, so another effect of taint is introduced into karmada. Users may taint the cluster by themselves manually. Meanwhile, `Failover` does depend on this feature, right?

Reply:
Oh, the feature gate should be added in #1781. Let me make a change here.

Reply:
Add a feature gate check in `cluster_controller`, it should  have been added in https://github.com/karmada-io/karmada/pull/1781. Thanks for reminding!

Reply:
> Actually, I think this feature is independent. We implement NoExecute taint manager, so another effect of taint is introduced into karmada. Users may taint the cluster by themselves manually. Meanwhile, Failover does depend on this feature, right?

After the eviction, the scheduler would select another cluster to fill the slot(if possible), which is exactly the scenario of Failover. For the case users taint the cluster, I think it's a special scenario that users initiate the `failover` by themselves. 

By the way, how do think about the `failover` feature? What are the use cases?

Reply:
> Oh, the feature gate should be added in https://github.com/karmada-io/karmada/pull/1781. Let me make a change here.

I'm ok with it. The feature gate means the feature is just introduced, maybe not mature enough to default enable, right?

Reply:
> By the way, how do think about the failover feature? What are the use cases?

1. Cluster has something wrong, ready condition becomes not ready or unknown.
2. Cluster controller tries to add NoExecute taint after 5 minutes, #1781.
3. Taint manager removes scheduling result of failed cluster in `rb.spec.clsuters`, #1945. It's a little bit like what `descheduler` does.
4. Scheduler tries to scale up these evicted replicas.

> I'm ok with it. The feature gate means the feature is just introduced, maybe not mature enough to default enable, right?

Not exactly, we just try to move `Failover` from scheduler to controller-manager.

Reply:
> Not exactly, we just try to move Failover from scheduler to controller-manager.

I see the plan at #1762. I think removing `failover` from the scheduler means relieving some responsibilities from the scheduler and letting the scheduler focus on `select` cluster for workload according to PropgationPolicy. Are we on the same page?

Anyway, let's have a talk at next week's meeting, are you available?

Reply:
Cool. No problem.